lib/update/cluster/phases/init.go

```diff
@@ -491,3 +503,126 @@ func (p *updatePhaseInit) removeConfiguredPackages() error {
 			return nil
 		})
 }
+
+func (r peerCleaner) cleanPeers() error {
+	if err := r.removeInvalidObjectPeerRefs(); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := r.removeLostObjects(); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := r.removeInvalidObjectPeers(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// removeLostObjects removes the object hashes from the state database
+// that only have peers that are no longer in the cluster
+func (r peerCleaner) removeLostObjects() error {
```

Comment:
Is there a chance it will nuke something it shouldn't? Is this a change that's necessary to recover some cluster? I'd be wary to deal directly with blobs in the object store esp. on every upgrade. What if there's a blob on a master node that doesn't have valid peers but is still used somehow?

Reply:
That's a good point and it's something I've been thinking with the alternative approach of adding TTL to a peer. We can treat the peer list in the object record as sort of history and ignore peers that aren't valid and, by extension, not attempting to pull objects with all invalid or empty peer list.